Core/clim-basic/clipboard.lisp

```diff
@@ -0,0 +1,174 @@
+
+(defgeneric clear-clipboard (port sheet)
+  (:documentation "Clear the content of the clipboard.
+
+When this function is called, the backend should ensure that the
+content of the selection or the clipboard is no longer available for
+pasting.
+
+PORT is the port that the sheet belongs to. It should be the primary
+argument that individual sheets specialise on.
+
+SHEET is the owner of the selection or clipboard.
+
+Clear the clipboard if CLIPBOARD-P is true, otherwise the selection.")
+  (:method ((port clim:port) (sheet clim:sheet))
+    nil))
+
+(defgeneric local-selection-content (port)
+  (:documentation "Returns the content of the selection in this Lisp image.
+If the global selection is currently held, the value returned is the
+same as what would be sent in response to a REQUEST-CLIPBOARD-CONTENT
+call.
+
+If content is available, this function should return two values: The
+content of the selection, and the presentation-type of the object.
+Otherwise NIL,NIL should be returned.
+
+This function is used by the copy selection to clipboard operation.
+This is in order to present the same beahviour as most non-CLIM
+applications exhibit: They are are able perform the copy
+operation (usually invoked by control-c) even when not owning the
+system-wide selection.
+
+The simplest implementation of this function would simply return the
+content of the selection, or NIL,NIL if it is not held.")
+  (:method ((port clim:port))
+    nil))
+
+;;;
+;;;  The following functions implement the standard API to request
+;;;  content from the clipboard.
+;;;
+
+(defgeneric request-clipboard-content (port pane type)
+  (:documentation "Backend implementation of REQUEST-CLIPBOARD-CONTENT.")
+  (:method ((port clim:port) (pane clim:sheet) type)
+    nil))
+
+(defgeneric request-selection-content (port pane type)
+  (:documentation "Request the content of the selection.
+
+Please refer to the documentation for REQUEST-CLIPBOARD-CONTENT for
+details.")
+  (:method ((port clim:port) (pane clim:sheet) type)
+    nil))
+
+(defmethod request-selection-content :around (port name type)
+  (unless (or (typep type 'climi::representation-type-name)
+              (and (listp type)
+                   (every (lambda (v) (typep v 'climi::representation-type-name)) type)))
+    (error "Invalid type: ~s" type))
+  (call-next-method))
+
+(defgeneric request-clipboard-content (port pane type)
+  (:documentation "Request the content of the clipboard.
+
+After calling this function, if the clipboard is active and can be
+presented as the requested type, an event of type CLIPBOARD-SEND-EVENT
+will be delivered to PANE. This event will contain the requested data
+in one of the requested formats.
+
+TYPE is a list of keywords (or a single keyword, in which case it is
+interpreted as a list of a single element) that indicates the type of
+data that can be accepted. If the clipboard content can be represented
+in more than one format, the first matching type will be used.")
```

Comment:
any keyword? can I put there ":i-like-pancakes"? if not, what are the values which may be put here?

Reply:
Yes. Any keyword can be used. Unsupported types are ignored.

Reply:
what's the point of not erroring on types which are *not* supported?

Reply:
The argument is that if you have a backend which does not, for example, support copying of `:STRING` objects, then you don't want to have an exception thrown whenever you type C-k in a Drei pane (which copies the string to the clipboard).

However, when calling either `REQUEST-*` or `COPY-*`, then you want to throw an exception if the caller specified some keyword other than the main ones supported (currently, `:STRING`, `:HTML` and `:IMAGE`).

You only want the backend to deal with the types it supports. The main check to verify that the caller used a proper type is verified upstream (in the `:AROUND` method).

Reply:
I still don't get why valid keywords are not specified given error is signalled in before method.

Reply:
As per previous discussion on IRC, I've removed the check.